docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,182 @@
+# Nullability API Design Notes
+
+## Guiding Principles
+
+When designing this API, there were a few core principles that we tried to keep in mind.
+
+1. First and foremost, we do not want to break existing analyzers and refactorings. This means that common patterns that used to work need to continue to work. These include reference equality for `ITypeSymbol`s that aren't generics, such as `String`.
+2. We'd like the ability to avoid computing nullability information if it's not required. Currently, the SemanticModel APIs do a good job of only causing the minimal amount of binding required to answer the question being asked, but attempting to compute nullability information will cause significantly more to need to be bound. To fully determine nullability, we must bind the entire method and run nullability analysis up to the point requested by the caller, unlike today where individual statements can get away with being bound by themselves. Even if the standard path calculates nullability info, we feel that some scenarios will be both perf sensitive and not care about results of nullability analysis. This includes things like intellisense.
+3. Users should be able to get at both declared nullability for variables, as well as seeing the current flow state of expressions based on flow analysis.
+
+## General Concepts
+
+Nullability information is exposed publicly as an enum, `Nullability`. There are 5 possible values:
```

Comment:
>5 possible values [](start = 81, length = 17)

If it's not covered below, it would be good to clarify what enum value is used for:
- a value type
- a nullable value type
- various generic type parameter scenarios